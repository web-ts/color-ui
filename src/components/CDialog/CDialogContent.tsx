import { Component, JSX, onMount, Signal } from "solid-js";

import { useDialogContent } from "@/composables/use-dialog-content";
import sleep from "@/utils/sleep";

import SCSS from "./CDialog.module.scss";

const candidateSelector = `input:not([inert]),select:not([inert]),textarea:not([inert]),a[href]:not([inert]),button:not([inert]),[tabindex]:not(slot):not([inert]),audio[controls]:not([inert]),video[controls]:not([inert]),[contenteditable]:not([contenteditable="false"]):not([inert]),details>summary:first-of-type:not([inert]),details:not([inert])`;

export const CDialogContent: Component<{
  model: Signal<boolean>;
  children: JSX.Element | JSX.Element[];
}> = (props) => {
  const dialogId = Math.random().toString(36).substring(7);

  const { inert } = useDialogContent(props.model, dialogId);

  let dialogRef: HTMLDivElement | undefined = undefined;

  onMount(async () => {
    const focusable = dialogRef?.querySelector(candidateSelector);
    if (!focusable) return;
    await sleep(10);
    (focusable as HTMLElement).focus?.();
  });

  return (
    <div
      ref={dialogRef}
      role="dialog"
      aria-modal="true"
      class={SCSS.CDialog}
      /* @ts-ignore inert is a valid html attribute*/
      inert={inert() ? "" : undefined}
    >
      {props.children}
    </div>
  );
};
