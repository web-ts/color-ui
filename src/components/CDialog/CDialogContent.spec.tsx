import { createSignal } from "solid-js";
import { describe, expect, it } from "vitest";

import sleep from "@/utils/sleep";
import { render } from "@solidjs/testing-library";

import { CDialogContent } from "./CDialogContent";

describe("CDialogContent", () => {
  it("should render children", () => {
    const signal = createSignal(true);
    const { container } = render(() => (
      <CDialogContent model={signal}>
        <div>Test</div>
      </CDialogContent>
    ));
    expect(container).toMatchSnapshot();
  });

  it("should focus if focusable element found", async () => {
    const signal = createSignal(true);
    render(() => (
      <CDialogContent model={signal}>
        <button id="test">Test</button>
      </CDialogContent>
    ));

    await sleep(11);

    console.log(document.activeElement?.id);

    expect(document.activeElement?.id).toBe("test");
  });
});
