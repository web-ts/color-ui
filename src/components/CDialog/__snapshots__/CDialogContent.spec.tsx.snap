// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CDialogContent > should render children 1`] = `
<div>
  <div
    aria-modal="true"
    class="_CDialog_45f97f"
    role="dialog"
  >
    <div>
      Test
    </div>
  </div>
</div>
`;
